File: 20230813_Sunday/20230813.rb
# Leetcode: 2369. Check if There is a Valid Partition For The Array.
# https://leetcode.com/problems/check-if-there-is-a-valid-partition-for-the-array/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 249 ms, faster than 100.00% of Ruby online submissions for Check if
# There is a Valid Partition For The Array.
# Memory Usage: 228.6 MB, less than 100.00% of Ruby online submissions for Check
# if There is a Valid Partition For The Array.
# 2023.08.13 Daily Challenge.
# @param {Integer[]} nums
# @return {Boolean}
def valid_partition(nums)
  # Based on:
  # https://leetcode.com/problems/check-if-there-is-a-valid-partition-for-the-array/solution/
  n = nums.size
  dp = Array.new(n+1,false)
  dp[0] = true
  n.times do |i|
    dpi = i + 1
    if i > 0 && nums[i] == nums[i-1]
      dp[dpi] ||= dp[dpi-2]
    end
    if i > 1 && nums[i] == nums[i-1] && nums[i-1] == nums[i-2]
      dp[dpi] ||= dp[dpi-3]
    end
    if i > 1 && nums[i] == nums[i-1] + 1 && nums[i-1] + 1 == nums[i-2]+2
      dp[dpi] ||=dp[dpi-3]
    end
  end
  return dp[n]
end

File: 20230813_Sunday/20230813.py
# Leetcode: 2369. Check if There is a Valid Partition For The Array.
# https://leetcode.com/problems/check-if-there-is-a-valid-partition-for-the-array/
class Solution:
    # Based on:
    # https://leetcode.com/problems/check-if-there-is-a-valid-partition-for-the-array/solution/
    def validPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        dp = [True] + [False] * n

        # Determine if the prefix array nums[0 ~ i] has a valid partition
        for i in range(n):
            dp_index = i + 1

            # Check 3 possibilities
            if i > 0 and nums[i] == nums[i - 1]:
                dp[dp_index] |= dp[dp_index - 2]
            if i > 1 and nums[i] == nums[i - 1] == nums[i - 2]:
                dp[dp_index] |= dp[dp_index - 3]
            if i > 1 and nums[i] == nums[i - 1] + 1 == nums[i - 2] + 2:
                dp[dp_index] |= dp[dp_index - 3]
 
        return dp[n]
